crypto/rsa/rsa_ossl.c

```diff
@@ -869,13 +901,8 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
          * verify will *always* be less than 'n' so we don't check for
          * absolute equality, just congruency.
          */
-        if (!BN_sub(vrfy, vrfy, I))
+        if (!BN_mod_sub_quick(vrfy, vrfy, I, rsa->n))
```

Comment:
Is this not just an equality check at this point? This code is checking that `vrfy` and `I` are equal mod `rsa->n`, but uses `BN_mod_sub_quick` which, as a precondition, requires that `vrfy` and `I` are fully reduced. That means that, if they were equal mod `rsa->n`, they're simply equal.

(Our version of this logic just does an equality check.)

Reply:
Commentary above suggests that I being larger than rsa->n is supposed to fly. As for question whether or not BN_mod_sub_quick does the job. Note that goal is not to produce real result modulo rsa->n, but to compare if verfy equals to I *or* I-rsa->n. Everything else, be it positive or negative, means that verification failed. I can add commentary...

Reply:
Hrm. If we allow `I` being larger than `rsa->n`, then the preconditions for the reduction in the smooth path don't hold and we have worse problems. It seems you'd want to fix up `I` earlier.

The callers of `rsa_mod_exp` within OpenSSL already check the input is fully-reduced (`RSA_R_DATA_TOO_LARGE_FOR_MODULUS`), so it is impossible there. But I suppose if I claim that `RSA_meth_get_mod_exp` is a concern above, I must also claim `RSA_meth_get_mod_exp` is a concern here. (Though it is slightly different. Saying you cannot use the result of `RSA_meth_get_mod_exp` at all is different from saying a mod-exp function had a previously unenforced requirement of reduced inputs.)

How much the project choses to care about `RSA_meth_get_mod_exp` I don't have opinions on. Personally I think it is a ridiculous function and should never have been included. :-)